src/storage/mvcc/txn.rs

```diff
@@ -1311,26 +1317,125 @@ mod tests {
         must_commit_err(&engine, k, 24, 25);
         must_rollback(&engine, k, 24);
 
+        // Acquire lock on a prewritten key should fail.
+        must_acquire_pessimistic_lock(&engine, k, k, 26, 26);
+        must_pessimistic_locked(&engine, k, 26, 26);
+        must_pessimistic_prewrite_delete(&engine, k, k, 26, 26, true);
+        must_locked(&engine, k, 26);
+        must_acquire_pessimistic_lock_err(&engine, k, k, 26, 26);
+        must_locked(&engine, k, 26);
+
+        // Acquire lock on a committed key should fail.
+        must_commit(&engine, k, 26, 27);
+        must_unlocked(&engine, k);
+        must_get_none(&engine, k, 28);
+        must_acquire_pessimistic_lock_err(&engine, k, k, 26, 26);
+        must_unlocked(&engine, k);
+        must_get_none(&engine, k, 28);
+        // Pessimistic prewrite on a committed key should fail.
+        must_pessimistic_prewrite_put_err(&engine, k, v, k, 26, 26, true);
+        must_unlocked(&engine, k);
+        must_get_none(&engine, k, 28);
+        // Currently we cannot avoid this.
+        must_acquire_pessimistic_lock(&engine, k, k, 26, 29);
+        must_pessimistic_rollback(&engine, k, 26, 29);
+        must_unlocked(&engine, k);
+
+        // Non pessimistic key in pessimistic transaction.
+        must_pessimistic_prewrite_put(&engine, k, v, k, 30, 30, false);
+        must_locked(&engine, k, 30);
+        must_commit(&engine, k, 30, 31);
+        must_unlocked(&engine, k);
+        must_get_commit_ts(&engine, k, 30, 31);
+
+        // Rollback collapsed.
+        must_rollback_collapsed(&engine, k, 32);
+        must_rollback_collapsed(&engine, k, 33);
+        must_acquire_pessimistic_lock_err(&engine, k, k, 32, 32);
+        // Currently we cannot avoid this.
```

Comment:
emmm, I don't think we can handle it since its rollback record has been removed. What we can do is pray there is no bug in TiDB.

Reply:
Basically, the rollback collapse logic assumes start ts must be greater than commit TS. 
Pessimistic transaction breaks the assumption.

Reply:
But if a pessimistic transaction has entered the prewrite phase, the forUpdateTS will not be updated anymore, so this is a impossible case.

Reply:
This case is possible when a `acquire_pessimistic_lock` request is delayed in the network and arrives after prewritting and rolling back the transaction. It's very unlikely to happen but possible.